dttools/src/list.c

```diff
@@ -1,406 +1,633 @@
+
+	return item;
 }
 
-void *list_peek_tail(struct list *l)
-{
-	if(!l)
-		return 0;
+void *list_peek_tail(struct list *l) {
+	void *item = NULL;
 
-	if(l->tail) {
-		return l->tail->data;
-	} else {
-		return 0;
-	}
+	if (!l)
+		return NULL;
+
+	struct list_cursor *cur = list_cursor_create(l);
+	list_seek(cur, -1);
+	list_get(cur, &item);
+	list_cursor_destroy(cur);
+
+	return item;
 }
 
-void *list_peek_current(struct list *l)
-{
-	if(!l)
-		return 0;
+void *list_peek_current(struct list *l) {
+	void *item = NULL;
 
-	if(l->iter) {
-		return l->iter->data;
-	} else {
-		return 0;
-	}
+	if (!l)
+		return NULL;
+
+	list_get(l->iter, &item);
+	return item;
 }
 
-void *list_remove(struct list *l, const void *value)
-{
-	struct list_node *n;
-	void *data;
+void *list_remove(struct list *l, const void *value) {
+	void *item;
+	void *out = NULL;
+	bool ok;
+
+	if (!value)
+		return NULL;
 
-	if(!value)
-		return 0;
-	for(n = l->head; n; n = n->next) {
-		if(value == n->data) {
-			data = n->data;
-			if(n->next)
-				n->next->prev = n->prev;
-			if(n->prev)
-				n->prev->next = n->next;
-			if(n == l->head)
-				l->head = n->next;
-			if(n == l->tail)
-				l->tail = n->prev;
-			free(n);
-			l->size--;
-			return data;
+	struct list_cursor *cur = list_cursor_create(l);
+	for (list_seek(cur, 0); list_get(cur, &item); list_next(cur)) {
+		if (value == item) {
+			out = item;
+			ok = list_drop(cur);
+			assert(ok);
+			break;
 		}
 	}
+	list_cursor_destroy(cur);
 
-	return 0;
+	return out;
 }
 
-void *list_find(struct list *l, list_op_t comparator, const void *arg)
-{
-	struct list_node *n;
+void *list_find(struct list *l, list_op_t comparator, const void *arg) {
+	void *item;
+	void *out = NULL;
 
-	for(n = l->head; n; n = n->next) {
-		if(comparator(n->data, arg))
-			return n->data;
+	struct list_cursor *cur = list_cursor_create(l);
+	for (list_seek(cur, 0); list_get(cur, &item); list_next(cur)) {
+		if(comparator(item, arg)) {
+			out = item;
+			break;
+		}
 	}
+	list_cursor_destroy(cur);
 
-	return 0;
+	return out;
 }
 
-int list_iterate(struct list *l, list_op_t operator, const void *arg)
-{
-	struct list_node *n;
+int list_iterate(struct list *l, list_op_t operator, const void *arg) {
+	void *item;
 	int alltheway = 1;
 
-	for(n = l->head; n; n = n->next) {
-		if(!operator(n->data, arg)) {
+	struct list_cursor *cur = list_cursor_create(l);
+	for (list_seek(cur, 0); list_get(cur, &item); list_next(cur)) {
+		if (!operator(item, arg)) {
 			alltheway = 0;
 			break;
 		}
 	}
+	list_cursor_destroy(cur);
 
 	return alltheway;
 }
 
-int list_iterate_reverse(struct list *l, list_op_t operator, const void *arg)
-{
-	struct list_node *n;
+int list_iterate_reverse(struct list *l, list_op_t operator, const void *arg) {
+	void *item;
 	int alltheway = 1;
 
-	for(n = l->tail; n; n = n->prev) {
-		if(!operator(n->data, arg)) {
+	struct list_cursor *cur = list_cursor_create(l);
+	for (list_seek(cur, -1); list_get(cur, &item); list_prev(cur)) {
```

Comment:
why -1?

Reply:
Iterating in reverse. Negative indices are taken from the back a la Python, so index -1 is the tail, and the cursor moves by `list_prev()`.